### PR TITLE
Allowed setting attributes on `DataLoader` and `BatchSampler` when instantiated inside `*_dataloader` hooks

### PR DESCRIPTION
**Pull request by otaj**
## What does this PR do?

Since #12981 and #13640, the new mechanism for re-instantiating `DataLoader` and `BatchSampler`s was adopted in case it was necessary to modify some of their behavior. While in general, this allowed for greater flexibility, it also introduced a subtle bug, that it was impossible to set attributes on these classes that would survive the re-instantiation process. To illustrate, let's take a look at the example below:

```python
loader = DataLoader(...)
loader.my_attribute = 10 # <- this will not survive re-instantiation
```

In order to fix it, this PR patches `__setattr__` and `__delattr__` methods, which serve as tape recorders, recording all the `__setattr__` and `__delattr__` calls happening on the object while in the proper context manager. During re-instantiation, all of these calls are replayed on top of the newly created object.

It also replaces all calls of `setattr(obj, <internal_pl_name>, value)` to `object.__setattr__(obj, <internal_pl_name>, value)` , especially within these contexts in order to not run into `RecursionError`s

It was discovered together with @krshrimali while debugging Flash CI

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @justusschock @awaelchli @ninginthecloud @rohitgr7 @otaj

### CODE REVIEW

**Comment by awaelchli** [comment]
This is a huge rabbit hole, what are we doing 🤣 🤣 
What's next? Let's also record all method calls on the dataloader and replaying them? They could have side effects outside the dataloader 🤯  Where will we stop? 😅

**Review by awaelchli — APPROVED** [review]

**Comment by otaj** [comment]
I fully agree, @awaelchli, that we're digging a rabbit hole here. And, to be fair, I was also thinking of recording the calls :joy: 

But, this was sadly a real issue, experienced by Flash and I'd rather have this solved for everyone and somewhat nicely, than telling @krshrimali hacks to make it work without this PR :joy:

**Review by krshrimali — APPROVED** [review]
Thanks, @otaj for the PR! 🚀 LGTM

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/utilities/data.py [hunk 1]
```diff
@@ -475,45 +503,82 @@ def wrapper(obj: Any, *args: Any, **kwargs: Any) -> None:
...
 
         if not hasattr(obj, "__pl_saved_args"):
-            obj.__pl_saved_args = args
-            obj.__pl_saved_kwargs = kwargs
-            obj.__pl_saved_arg_names = param_names
-            obj.__pl_saved_default_kwargs = default_kwargs
+            object.__setattr__(obj, "__pl_saved_args", args)
+            object.__setattr__(obj, "__pl_saved_kwargs", kwargs)
+            object.__setattr__(obj, "__pl_saved_arg_names", param_names)
+            object.__setattr__(obj, "__pl_saved_default_kwargs", default_kwargs)
```
Comment (justusschock) [inline src/pytorch_lightning/utilities/data.py:509]: do we need to call this ecplicitly? can't we do `setattr(obj, "__pl_saved_args", args)`

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/utilities/data.py (hunk 1 above)
Comment (otaj) [reply src/pytorch_lightning/utilities/data.py:509]: I'd almost say that's a philosophical question :joy:

If for some reason a user overwrites their `__setattr__` (for example to mimic frozen class), then calling normal `setattr(obj, ..., ...)` would not work. So, that and running into `RecursionError`s while testing were the reasons why I decided to use `object.__setattr__(obj, ..., ...)` everywhere in the case of these internal attributes

Comment (justusschock) [reply src/pytorch_lightning/utilities/data.py:509]: Makes sense, I was just wondering if it was an issue with class binding or whether this was for other reasons :) Because I actually prefer not to call magic methods directly if possible, but in this case it makes sense :)

File: src/pytorch_lightning/utilities/data.py [hunk 2]
```diff
@@ -49,6 +49,18 @@
...
 
 
+class _WrapAttrTag(LightningEnum):
+    SET = "set"
+    DEL = "del"
+
+    def __call__(self, *args):
+        if self == self.SET:
+            fn = setattr
+        else:
```
Comment (Borda) [inline src/pytorch_lightning/utilities/data.py:59]: elif?

File: src/pytorch_lightning/utilities/data.py [hunk 3]
```diff
@@ -450,6 +442,40 @@ def _auto_add_worker_init_fn(dataloader: DataLoader, rank: int) -> None:
         dataloader.worker_init_fn = partial(pl_worker_init_function, rank=rank)
 
 
+def _reinstantiate_wrapped_cls(orig_object: Any, *args: Any, explicit_cls: Optional[Type] = None, **kwargs: Any) -> Any:
+    if explicit_cls is None:
+        constructor = type(orig_object)
+    else:
+        constructor = explicit_cls
```
Comment (Borda) [inline src/pytorch_lightning/utilities/data.py:449]:
```suggestion
    constructor = type(orig_object) if explicit_cls is None else explicit_cls
```

**Review by Borda — APPROVED** [review]
stretched my knowledge a bit, but so far I can say, LGTM

**Review by carmocca — APPROVED** [review]
Did not review. Will check it in the future. Approving to unblock 👍

File: src/pytorch_lightning/utilities/data.py (hunk 2 above)
Comment (otaj) [reply src/pytorch_lightning/utilities/data.py:59]: Not needed since it's an `Enum` with only two possible values. It cannot be constructed with any other value
